Core/include/Acts/Vertexing/AdaptiveMultiVertexFitter.ipp

```diff
@@ -188,25 +181,42 @@ bool Acts::AdaptiveMultiVertexFitter<input_track_t, linearizer_t>::
 }
 
 template <typename input_track_t, typename linearizer_t>
-Acts::Result<void> Acts::
-    AdaptiveMultiVertexFitter<input_track_t, linearizer_t>::prepareVertexForFit(
+Acts::Result<void>
+Acts::AdaptiveMultiVertexFitter<input_track_t, linearizer_t>::
+    updateImpactParams3D(
         State& state, Vertex<input_track_t>* vtx,
         const VertexingOptions<input_track_t>& vertexingOptions) const {
   // The current vertex info object
   auto& currentVtxInfo = state.vtxInfoMap[vtx];
-  // The seed position
-  const Vector3& seedPos = currentVtxInfo.seedPosition.template head<3>();
+  // Vertex position, i.e., point wrt which the impact parameters are estimated
+  const Vector3& vtxPosition = currentVtxInfo.oldPosition.template head<3>();
 
   // Loop over all tracks at current vertex
   for (const auto& trk : currentVtxInfo.trackLinks) {
+    // Boolean indicating whether the impact parameters were already calculated
+    bool hasIP = currentVtxInfo.impactParams3D.count(trk) == 1;
+    // Track parameters
+    auto trkParams = m_extractParameters(*trk);
+    // Origin of the track reference surface
+    Vector3 surfaceOrigin = trkParams.referenceSurface()
+                                .transform(vertexingOptions.geoContext)
+                                .translation();
+    // Skip the impact point estimation if the impact parameters of trk wrt the
+    // vertex position were already calculated
+    if (hasIP and surfaceOrigin == vtxPosition) {
+      continue;
+    }
     auto res = m_cfg.ipEst.estimate3DImpactParameters(
         vertexingOptions.geoContext, vertexingOptions.magFieldContext,
-        m_extractParameters(*trk), seedPos, state.ipState);
+        trkParams, vtxPosition, state.ipState);
     if (!res.ok()) {
       return res.error();
     }
-    // Set ip3dParams for current trackAtVertex
-    currentVtxInfo.ip3dParams.emplace(trk, res.value());
+    if (hasIP) {
+      currentVtxInfo.impactParams3D.at(trk) = res.value();
+    } else {
+      currentVtxInfo.impactParams3D.emplace(trk, res.value());
+    }
```

Comment:
I don't think you need to do this, `at` should auto-create an element if it doesn't exist. If not, I'm pretty sure `operator[]` does it in any case.